File: water_counter.go
# water_counter.go
/**
 * 💧 Water Counter – Daily Hydration Tracker (Go Edition)
 * Advanced: colored output, JSON persistence, analytics, export/import
 */

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

// ─── Types ────────────────────────────────────────────────────────────────────

type WaterEntry struct {
	Date      string `json:"date"`
	Amount    int    `json:"amount"`
	Timestamp string `json:"timestamp"`
}

type WaterData struct {
	Goal    int           `json:"goal"`
	Entries []WaterEntry `json:"entries"`
}

// ─── Constants ────────────────────────────────────────────────────────────────

const (
	defaultGoal = 2000
	maxAmount   = 10000
	historyDays = 7
)

// ─── Colors ──────────────────────────────────────────────────────────────────

const (
	reset  = "\x1b[0m"
	bright = "\x1b[1m"
	dim    = "\x1b[2m"
	red    = "\x1b[31m"
	green  = "\x1b[32m"
	yellow = "\x1b[33m"
	blue   = "\x1b[34m"
	magenta = "\x1b[35m"
	cyan   = "\x1b[36m"
)

func c(str, color string) string {
	return color + str + reset
}

// ─── Main App ────────────────────────────────────────────────────────────────

type WaterCounter struct {
	goal      int
	entries   []WaterEntry
	dataPath  string
	reader    *bufio.Reader
}

func NewWaterCounter() *WaterCounter {
	home, err := os.UserHomeDir()
	if err != nil {
		panic(err)
	}
	dataDir := filepath.Join(home, ".water_counter")
	if err := os.MkdirAll(dataDir, 0755); err != nil {
		panic(err)
	}
	dataPath := filepath.Join(dataDir, "data.json")

	wc := &WaterCounter{
		dataPath: dataPath,
		reader:   bufio.NewReader(os.Stdin),
	}
	wc.loadData()
	return wc
}

// ─── Data Persistence ──────────────────────────────────────────────────────

func (w *WaterCounter) loadData() {
	if _, err := os.Stat(w.dataPath); os.IsNotExist(err) {
		w.goal = defaultGoal
		w.entries = []WaterEntry{}
		return
	}

	raw, err := ioutil.ReadFile(w.dataPath)
	if err != nil {
		fmt.Println(c("⚠️  Error loading data. Starting fresh.", yellow))
		w.goal = defaultGoal
		w.entries = []WaterEntry{}
		return
	}

	var data WaterData
	if err := json.Unmarshal(raw, &data); err != nil {
		fmt.Println(c("⚠️  Error parsing data. Starting fresh.", yellow))
		w.goal = defaultGoal
		w.entries = []WaterEntry{}
		return
	}

	w.goal = data.Goal
	if w.goal <= 0 {
		w.goal = defaultGoal
	}
	w.entries = data.Entries
	if w.entries == nil {
		w.entries = []WaterEntry{}
	}
}

func (w *WaterCounter) saveData() error {
	data := WaterData{
		Goal:    w.goal,
		Entries: w.entries,
	}
	raw, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	tempPath := w.dataPath + ".tmp"
	if err := ioutil.WriteFile(tempPath, raw, 0644); err != nil {
		return err
	}
	return os.Rename(tempPath, w.dataPath)
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

func (w *WaterCounter) today() string {
	return time.Now().Format("2006-01-02")
}

func (w *WaterCounter) getTodayEntries() []WaterEntry {
	today := w.today()
	var result []WaterEntry
	for _, e := range w.entries {
		if e.Date == today {
			result = append(result, e)
		}
	}
	return result
}

func (w *WaterCounter) getTodayTotal() int {
	total := 0
	for _, e := range w.getTodayEntries() {
		total += e.Amount
	}
	return total
}

func (w *WaterCounter) getProgressBar(current, goal, width int) string {
	if goal <= 0 {
		return "⚠️  Goal not set"
	}
	ratio := float64(current) / float64(goal)
	if ratio > 1.0 {
		ratio = 1.0
	}
	filled := int(ratio * float64(width))
	bar := strings.Repeat("█", filled) + strings.Repeat("░", width-filled)
	return fmt.Sprintf("%s %.1f%%", bar, ratio*100)
}

func (w *WaterCounter) ask(prompt string) string {
	fmt.Print(prompt)
	line, _ := w.reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func (w *WaterCounter) askInt(prompt string) (int, error) {
	answer := w.ask(prompt)
	return strconv.Atoi(answer)
}

func (w *WaterCounter) askConfirm(prompt string) bool {
	answer := w.ask(prompt + " (yes/no): ")
	answer = strings.ToLower(answer)
	return answer == "yes" || answer == "y"
}

// ─── Core Actions ──────────────────────────────────────────────────────────

func (w *WaterCounter) AddEntry(amount int, date ...string) error {
	if amount <= 0 {
		return fmt.Errorf("amount must be positive")
	}
	if amount > maxAmount {
		fmt.Println(c(fmt.Sprintf("⚠️  That's a LOT of water! Max %dml", maxAmount), yellow))
		if !w.askConfirm("Continue?") {
			return nil
		}
	}

	entryDate := w.today()
	if len(date) > 0 && date[0] != "" {
		entryDate = date[0]
	}
	timestamp := time.Now().Format(time.RFC3339)

	w.entries = append(w.entries, WaterEntry{
		Date:      entryDate,
		Amount:    amount,
		Timestamp: timestamp,
	})
	if err := w.saveData(); err != nil {
		return err
	}

	todayTotal := w.getTodayTotal()
	fmt.Printf(c("✅ Added %dml (Total today: %dml)\n", green), amount, todayTotal)
	if todayTotal >= w.goal {
		fmt.Println(c("🎉 Goal achieved! Stay hydrated! 💪", cyan))
	}
	return nil
}

func (w *WaterCounter) ShowToday() {
	todayTotal := w.getTodayTotal()
	entries := w.getTodayEntries()

	fmt.Println("\n" + c(strings.Repeat("═", 50), dim))
	fmt.Println(c("💧 TODAY'S HYDRATION", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), dim))
	fmt.Printf("  Goal:      %s\n", c(fmt.Sprintf("%dml", w.goal), cyan))
	fmt.Printf("  Consumed:  %s\n", c(fmt.Sprintf("%dml", todayTotal), green))
	fmt.Printf("  Remaining: %s\n", c(fmt.Sprintf("%dml", max(w.goal-todayTotal, 0)), yellow))
	fmt.Printf("  Progress:  %s\n", w.getProgressBar(todayTotal, w.goal, 30))
	fmt.Println(c(strings.Repeat("═", 50), dim))

	if len(entries) > 0 {
		fmt.Println("  Entries:")
		for i, e := range entries {
			ts := "—"
			if len(e.Timestamp) >= 16 {
				ts = e.Timestamp[11:16]
			}
			fmt.Printf("    %d. %s → %s\n", i+1, ts, c(fmt.Sprintf("%dml", e.Amount), green))
		}
	} else {
		fmt.Println(c("  No entries yet today. Drink up! 💧", dim))
	}
	fmt.Println(c(strings.Repeat("═", 50), dim))
}

func (w *WaterCounter) ShowHistory(days ...int) {
	n := historyDays
	if len(days) > 0 && days[0] > 0 {
		n = days[0]
	}

	now := time.Now()
	history := make([]struct {
		Date  string
		Total int
	}, n)

	for i := 0; i < n; i++ {
		d := now.AddDate(0, 0, -i)
		dateStr := d.Format("2006-01-02")
		total := 0
		for _, e := range w.entries {
			if e.Date == dateStr {
				total += e.Amount
			}
		}
		history[i] = struct {
			Date  string
			Total int
		}{Date: dateStr, Total: total}
	}

	fmt.Printf("\n📅 LAST %d DAYS\n", n)
	fmt.Println(c(strings.Repeat("─", 40), dim))
	for _, h := range history {
		status := "❌"
		if h.Total >= w.goal {
			status = "✅"
		} else if h.Total > 0 {
			status = "⏳"
		}
		fmt.Printf("  %s: %5dml  %s\n", h.Date, h.Total, status)
	}
	fmt.Println(c(strings.Repeat("─", 40), dim))
}

func (w *WaterCounter) ShowStats() {
	if len(w.entries) == 0 {
		fmt.Println(c("📭 No data yet. Start tracking your water intake!", yellow))
		return
	}

	total := 0
	minEntry := int(^uint(0) >> 1)
	maxEntry := 0
	for _, e := range w.entries {
		total += e.Amount
		if e.Amount < minEntry {
			minEntry = e.Amount
		}
		if e.Amount > maxEntry {
			maxEntry = e.Amount
		}
	}
	count := len(w.entries)
	avg := float64(total) / float64(count)

	uniqueDays := make(map[string]bool)
	for _, e := range w.entries {
		uniqueDays[e.Date] = true
	}
	dayCount := len(uniqueDays)

	// Best day
	dayTotals := make(map[string]int)
	for _, e := range w.entries {
		dayTotals[e.Date] += e.Amount
	}
	bestDay := ""
	bestAmount := 0
	for date, amt := range dayTotals {
		if amt > bestAmount {
			bestAmount = amt
			bestDay = date
		}
	}

	fmt.Println("\n" + c(strings.Repeat("═", 50), dim))
	fmt.Println(c("📊 STATISTICS", bright+magenta))
	fmt.Println(c(strings.Repeat("═", 50), dim))
	fmt.Printf("  Total consumed:  %s\n", c(fmt.Sprintf("%dml", total), cyan))
	fmt.Printf("  Total entries:   %s\n", c(fmt.Sprintf("%d", count), cyan))
	fmt.Printf("  Days tracked:    %s\n", c(fmt.Sprintf("%d", dayCount), cyan))
	fmt.Printf("  Average per day: %s\n", c(fmt.Sprintf("%.1fml", avg), cyan))
	fmt.Printf("  Max entry:       %s\n", c(fmt.Sprintf("%dml", maxEntry), green))
	fmt.Printf("  Min entry:       %s\n", c(fmt.Sprintf("%dml", minEntry), yellow))
	bestStr := "—"
	if bestDay != "" {
		bestStr = fmt.Sprintf("%s (%dml)", bestDay, bestAmount)
	}
	fmt.Printf("  Best day:        %s\n", c(bestStr, green))
	fmt.Printf("  Daily goal:      %s\n", c(fmt.Sprintf("%dml", w.goal), cyan))
	fmt.Println(c(strings.Repeat("═", 50), dim))
}

func (w *WaterCounter) SetGoal(goal int) error {
	if goal <= 0 {
		return fmt.Errorf("goal must be positive")
	}
	if goal > maxAmount {
		fmt.Println(c(fmt.Sprintf("⚠️  That's an extreme goal! Max %dml", maxAmount), yellow))
		if !w.askConfirm("Set anyway?") {
			return nil
		}
	}
	w.goal = goal
	if err := w.saveData(); err != nil {
		return err
	}
	fmt.Printf(c("✅ Daily goal set to %dml\n", green), goal)
	return nil
}

func (w *WaterCounter) ExportData(filepath ...string) error {
	fp := ""
	if len(filepath) > 0 && filepath[0] != "" {
		fp = filepath[0]
	} else {
		ts := time.Now().Format("20060102_150405")
		fp = fmt.Sprintf("water_counter_export_%s.json", ts)
	}

	data := WaterData{Goal: w.goal, Entries: w.entries}
	raw, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	if err := ioutil.WriteFile(fp, raw, 0644); err != nil {
		return err
	}
	fmt.Printf(c("✅ Data exported to: %s\n", green), fp)
	return nil
}

func (w *WaterCounter) ImportData(filepath string) error {
	if _, err := os.Stat(filepath); os.IsNotExist(err) {
		return fmt.Errorf("file not found: %s", filepath)
	}
	raw, err := ioutil.ReadFile(filepath)
	if err != nil {
		return err
	}
	var data WaterData
	if err := json.Unmarshal(raw, &data); err != nil {
		return fmt.Errorf("invalid data format: %v", err)
	}
	if data.Goal <= 0 || data.Entries == nil {
		return fmt.Errorf("invalid data format")
	}
	if !w.askConfirm("⚠️  This will overwrite current data! Continue?") {
		return nil
	}
	w.goal = data.Goal
	w.entries = data.Entries
	if err := w.saveData(); err != nil {
		return err
	}
	fmt.Printf(c("✅ Imported %d entries from: %s\n", green), len(w.entries), filepath)
	return nil
}

func (w *WaterCounter) ClearData() error {
	if len(w.entries) == 0 && w.goal == defaultGoal {
		fmt.Println(c("📭 Already empty.", dim))
		return nil
	}
	if !w.askConfirm("⚠️  Delete ALL data? This cannot be undone!") {
		return nil
	}
	w.entries = []WaterEntry{}
	w.goal = defaultGoal
	if err := w.saveData(); err != nil {
		return err
	}
	fmt.Println(c("🗑️  All data cleared.", yellow))
	return nil
}

// ─── Menu ────────────────────────────────────────────────────────────────────

func (w *WaterCounter) showMenu() {
	todayTotal := w.getTodayTotal()
	progress := w.getProgressBar(todayTotal, w.goal, 30)

	fmt.Println("\n" + c(strings.Repeat("═", 50), cyan))
	fmt.Println(c("💧 WATER COUNTER – Daily Hydration Tracker", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Printf("  Today: %dml / %dml  %s\n", todayTotal, w.goal, progress)
	fmt.Println(c(strings.Repeat("─", 50), dim))
	fmt.Println("  1. 💧 Add water intake")
	fmt.Println("  2. 📊 Show today's progress")
	fmt.Printf("  3. 📅 Show history (%d days)\n", historyDays)
	fmt.Println("  4. 📈 Show statistics")
	fmt.Printf("  5. 🎯 Set daily goal (current: %dml)\n", w.goal)
	fmt.Println("  6. 📤 Import / Export data")
	fmt.Println("  7. 🗑️  Clear all data")
	fmt.Println("  0. 🚪 Exit")
	fmt.Println(c(strings.Repeat("═", 50), cyan))
}

func (w *WaterCounter) handleImportExport() {
	fmt.Println("\n📤 IMPORT / EXPORT")
	fmt.Println("  1. Export data")
	fmt.Println("  2. Import data")
	fmt.Println("  0. Back")
	choice := w.ask("Your choice: ")

	switch choice {
	case "1":
		if err := w.ExportData(); err != nil {
			fmt.Println(c("❌ Export failed: "+err.Error(), red))
		}
	case "2":
		fp := w.ask("Path to JSON file: ")
		if fp != "" {
			if err := w.ImportData(fp); err != nil {
				fmt.Println(c("❌ Import failed: "+err.Error(), red))
			}
		}
	}
}

func (w *WaterCounter) Run() {
	fmt.Print("\033[H\033[2J")
	fmt.Println(c("\n💧 Water Counter – Daily Hydration Tracker", bright+cyan))
	fmt.Println(c("Stay hydrated, stay healthy!", dim))

	for {
		w.showMenu()
		choice := w.ask("Your choice: ")

		switch choice {
		case "1":
			amount, err := w.askInt("Amount in ml: ")
			if err == nil {
				if err := w.AddEntry(amount); err != nil {
					fmt.Println(c("❌ "+err.Error(), red))
				}
			} else {
				fmt.Println(c("❌ Please enter a number.", red))
			}
		case "2":
			w.ShowToday()
		case "3":
			w.ShowHistory()
		case "4":
			w.ShowStats()
		case "5":
			goal, err := w.askInt("New daily goal (ml): ")
			if err == nil {
				if err := w.SetGoal(goal); err != nil {
					fmt.Println(c("❌ "+err.Error(), red))
				}
			} else {
				fmt.Println(c("❌ Please enter a number.", red))
			}
		case "6":
			w.handleImportExport()
		case "7":
			if err := w.ClearData(); err != nil {
				fmt.Println(c("❌ "+err.Error(), red))
			}
		case "0":
			fmt.Println(c("👋 Stay hydrated! Goodbye!", cyan))
			return
		default:
			fmt.Println(c("❌ Invalid choice.", red))
		}

		if choice != "0" {
			fmt.Print("\nPress Enter to continue...")
			w.reader.ReadString('\n')
		}
	}
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// ─── Main ─────────────────────────────────────────────────────────────────────

func main() {
	app := NewWaterCounter()
	app.Run()
}

File: water_counter.rs
# water_counter.rs
/**
 * 💧 Water Counter – Daily Hydration Tracker (Rust Edition)
 * Advanced: colorful terminal, JSON persistence, analytics, export/import
 */

use chrono::{DateTime, Local, NaiveDate};
use serde::{Deserialize, Serialize};
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::PathBuf;

// ─── Types ──────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
struct WaterEntry {
    date: String,
    amount: u32,
    timestamp: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct WaterData {
    goal: u32,
    entries: Vec<WaterEntry>,
}

// ─── Constants ─────────────────────────────────────────────────────────────

const DEFAULT_GOAL: u32 = 2000;
const MAX_AMOUNT: u32 = 10000;
const HISTORY_DAYS: usize = 7;

// ─── Colors ──────────────────────────────────────────────────────────────────

fn c(text: &str, color: &str) -> String {
    format!("{}{}{}", color, text, "\x1b[0m")
}

const RESET: &str = "\x1b[0m";
const BRIGHT: &str = "\x1b[1m";
const DIM: &str = "\x1b[2m";
const RED: &str = "\x1b[31m";
const GREEN: &str = "\x1b[32m";
const YELLOW: &str = "\x1b[33m";
const BLUE: &str = "\x1b[34m";
const MAGENTA: &str = "\x1b[35m";
const CYAN: &str = "\x1b[36m";

// ─── Main App ──────────────────────────────────────────────────────────────

struct WaterCounter {
    goal: u32,
    entries: Vec<WaterEntry>,
    data_path: PathBuf,
}

impl WaterCounter {
    fn new() -> Result<Self, Box<dyn std::error::Error>> {
        let home = std::env::var("HOME").or_else(|_| std::env::var("USERPROFILE"))?;
        let mut data_dir = PathBuf::from(home);
        data_dir.push(".water_counter");
        fs::create_dir_all(&data_dir)?;
        let data_path = data_dir.join("data.json");

        let mut app = WaterCounter {
            goal: DEFAULT_GOAL,
            entries: Vec::new(),
            data_path,
        };
        app.load_data()?;
        Ok(app)
    }

    // ─── Data Persistence ──────────────────────────────────────────────────

    fn load_data(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        if !self.data_path.exists() {
            self.goal = DEFAULT_GOAL;
            self.entries = Vec::new();
            return Ok(());
        }

        let raw = fs::read_to_string(&self.data_path)?;
        match serde_json::from_str::<WaterData>(&raw) {
            Ok(data) => {
                self.goal = if data.goal > 0 { data.goal } else { DEFAULT_GOAL };
                self.entries = data.entries;
                Ok(())
            }
            Err(e) => {
                eprintln!("{}", c("⚠️  Error parsing data. Starting fresh.", YELLOW));
                self.goal = DEFAULT_GOAL;
                self.entries = Vec::new();
                Ok(())
            }
        }
    }

    fn save_data(&self) -> Result<(), Box<dyn std::error::Error>> {
        let data = WaterData {
            goal: self.goal,
            entries: self.entries.clone(),
        };
        let raw = serde_json::to_string_pretty(&data)?;
        let temp_path = self.data_path.with_extension("tmp");
        fs::write(&temp_path, raw)?;
        fs::rename(&temp_path, &self.data_path)?;
        Ok(())
    }

    // ─── Helpers ──────────────────────────────────────────────────────────

    fn today(&self) -> String {
        Local::now().format("%Y-%m-%d").to_string()
    }

    fn get_today_entries(&self) -> Vec<WaterEntry> {
        let today = self.today();
        self.entries
            .iter()
            .filter(|e| e.date == today)
            .cloned()
            .collect()
    }

    fn get_today_total(&self) -> u32 {
        self.get_today_entries().iter().map(|e| e.amount).sum()
    }

    fn get_progress_bar(&self, current: u32, goal: u32, width: usize) -> String {
        if goal == 0 {
            return "⚠️  Goal not set".to_string();
        }
        let ratio = (current as f64 / goal as f64).min(1.0);
        let filled = (ratio * width as f64) as usize;
        let bar = "█".repeat(filled) + &"░".repeat(width - filled);
        format!("{} {:.1}%", bar, ratio * 100.0)
    }

    fn ask(prompt: &str) -> String {
        print!("{}", prompt);
        io::stdout().flush().unwrap();
        let stdin = io::stdin();
        let mut line = String::new();
        stdin.lock().read_line(&mut line).unwrap();
        line.trim().to_string()
    }

    fn ask_int(prompt: &str) -> Result<u32, Box<dyn std::error::Error>> {
        let answer = Self::ask(prompt);
        Ok(answer.parse()?)
    }

    fn ask_confirm(prompt: &str) -> bool {
        let answer = Self::ask(&format!("{} (yes/no): ", prompt));
        let a = answer.to_lowercase();
        a == "yes" || a == "y"
    }

    // ─── Core Actions ──────────────────────────────────────────────────────

    fn add_entry(&mut self, amount: u32) -> Result<(), Box<dyn std::error::Error>> {
        if amount == 0 {
            return Err("Amount must be positive!".into());
        }
        if amount > MAX_AMOUNT {
            eprintln!("{}", c(&format!("⚠️  That's a LOT of water! Max {}ml", MAX_AMOUNT), YELLOW));
            if !Self::ask_confirm("Continue?") {
                return Ok(());
            }
        }

        let date = self.today();
        let timestamp = Local::now().to_rfc3339();

        self.entries.push(WaterEntry {
            date,
            amount,
            timestamp,
        });
        self.save_data()?;

        let today_total = self.get_today_total();
        println!("{}", c(&format!("✅ Added {}ml (Total today: {}ml)", amount, today_total), GREEN));
        if today_total >= self.goal {
            println!("{}", c("🎉 Goal achieved! Stay hydrated! 💪", CYAN));
        }
        Ok(())
    }

    fn show_today(&self) {
        let today_total = self.get_today_total();
        let entries = self.get_today_entries();

        println!("\n{}", c(&"═".repeat(50), DIM));
        println!("{}", c("💧 TODAY'S HYDRATION", &format!("{}{}", BRIGHT, CYAN)));
        println!("{}", c(&"═".repeat(50), DIM));
        println!("  Goal:      {}", c(&format!("{}ml", self.goal), CYAN));
        println!("  Consumed:  {}", c(&format!("{}ml", today_total), GREEN));
        println!("  Remaining: {}", c(&format!("{}ml", self.goal.saturating_sub(today_total)), YELLOW));
        println!("  Progress:  {}", self.get_progress_bar(today_total, self.goal, 30));
        println!("{}", c(&"═".repeat(50), DIM));

        if entries.is_empty() {
            println!("{}", c("  No entries yet today. Drink up! 💧", DIM));
        } else {
            println!("  Entries:");
            for (i, e) in entries.iter().enumerate() {
                let ts = if e.timestamp.len() >= 16 {
                    &e.timestamp[11..16]
                } else {
                    "—"
                };
                println!("    {}. {} → {}", i + 1, ts, c(&format!("{}ml", e.amount), GREEN));
            }
        }
        println!("{}", c(&"═".repeat(50), DIM));
    }

    fn show_history(&self, days: usize) {
        let now = Local::now();
        let mut history = Vec::new();

        for i in 0..days {
            let d = now - chrono::Duration::days(i as i64);
            let date_str = d.format("%Y-%m-%d").to_string();
            let total: u32 = self.entries
                .iter()
                .filter(|e| e.date == date_str)
                .map(|e| e.amount)
                .sum();
            history.push((date_str, total));
        }

        println!("\n📅 LAST {} DAYS", days);
        println!("{}", c(&"─".repeat(40), DIM));
        for (date, total) in history {
            let status = if total >= self.goal {
                "✅"
            } else if total > 0 {
                "⏳"
            } else {
                "❌"
            };
            println!("  {}: {:>5}ml  {}", date, total, status);
        }
        println!("{}", c(&"─".repeat(40), DIM));
    }

    fn show_stats(&self) {
        if self.entries.is_empty() {
            println!("{}", c("📭 No data yet. Start tracking your water intake!", YELLOW));
            return;
        }

        let total: u32 = self.entries.iter().map(|e| e.amount).sum();
        let count = self.entries.len();
        let avg = total as f64 / count as f64;

        let max_entry = self.entries.iter().map(|e| e.amount).max().unwrap_or(0);
        let min_entry = self.entries.iter().map(|e| e.amount).min().unwrap_or(0);

        let mut unique_days = std::collections::HashSet::new();
        for e in &self.entries {
            unique_days.insert(&e.date);
        }
        let day_count = unique_days.len();

        // Best day
        let mut day_totals = std::collections::HashMap::new();
        for e in &self.entries {
            *day_totals.entry(&e.date).or_insert(0) += e.amount;
        }
        let mut best_day = "";
        let mut best_amount = 0;
        for (date, amt) in &day_totals {
            if *amt > best_amount {
                best_amount = *amt;
                best_day = date;
            }
        }

        println!("\n{}", c(&"═".repeat(50), DIM));
        println!("{}", c("📊 STATISTICS", &format!("{}{}", BRIGHT, MAGENTA)));
        println!("{}", c(&"═".repeat(50), DIM));
        println!("  Total consumed:  {}", c(&format!("{}ml", total), CYAN));
        println!("  Total entries:   {}", c(&format!("{}", count), CYAN));
        println!("  Days tracked:    {}", c(&format!("{}", day_count), CYAN));
        println!("  Average per day: {}", c(&format!("{:.1}ml", avg), CYAN));
        println!("  Max entry:       {}", c(&format!("{}ml", max_entry), GREEN));
        println!("  Min entry:       {}", c(&format!("{}ml", min_entry), YELLOW));
        let best_str = if best_day.is_empty() {
            "—".to_string()
        } else {
            format!("{} ({}ml)", best_day, best_amount)
        };
        println!("  Best day:        {}", c(&best_str, GREEN));
        println!("  Daily goal:      {}", c(&format!("{}ml", self.goal), CYAN));
        println!("{}", c(&"═".repeat(50), DIM));
    }

    fn set_goal(&mut self, goal: u32) -> Result<(), Box<dyn std::error::Error>> {
        if goal == 0 {
            return Err("Goal must be positive!".into());
        }
        if goal > MAX_AMOUNT {
            eprintln!("{}", c(&format!("⚠️  That's an extreme goal! Max {}ml", MAX_AMOUNT), YELLOW));
            if !Self::ask_confirm("Set anyway?") {
                return Ok(());
            }
        }
        self.goal = goal;
        self.save_data()?;
        println!("{}", c(&format!("✅ Daily goal set to {}ml", goal), GREEN));
        Ok(())
    }

    fn export_data(&self, filepath: Option<&str>) -> Result<(), Box<dyn std::error::Error>> {
        let fp = if let Some(p) = filepath {
            p.to_string()
        } else {
            let ts = Local::now().format("%Y%m%d_%H%M%S").to_string();
            format!("water_counter_export_{}.json", ts)
        };

        let data = WaterData {
            goal: self.goal,
            entries: self.entries.clone(),
        };
        let raw = serde_json::to_string_pretty(&data)?;
        fs::write(&fp, raw)?;
        println!("{}", c(&format!("✅ Data exported to: {}", fp), GREEN));
        Ok(())
    }

    fn import_data(&mut self, filepath: &str) -> Result<(), Box<dyn std::error::Error>> {
        if !std::path::Path::new(filepath).exists() {
            return Err(format!("File not found: {}", filepath).into());
        }
        let raw = fs::read_to_string(filepath)?;
        let data: WaterData = serde_json::from_str(&raw)?;
        if data.goal == 0 || data.entries.is_empty() {
            return Err("Invalid data format!".into());
        }
        if !Self::ask_confirm("⚠️  This will overwrite current data! Continue?") {
            return Ok(());
        }
        self.goal = data.goal;
        self.entries = data.entries;
        self.save_data()?;
        println!("{}", c(&format!("✅ Imported {} entries from: {}", self.entries.len(), filepath), GREEN));
        Ok(())
    }

    fn clear_data(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        if self.entries.is_empty() && self.goal == DEFAULT_GOAL {
            println!("{}", c("📭 Already empty.", DIM));
            return Ok(());
        }
        if !Self::ask_confirm("⚠️  Delete ALL data? This cannot be undone!") {
            return Ok(());
        }
        self.entries.clear();
        self.goal = DEFAULT_GOAL;
        self.save_data()?;
        println!("{}", c("🗑️  All data cleared.", YELLOW));
        Ok(())
    }

    // ─── Menu ──────────────────────────────────────────────────────────────

    fn show_menu(&self) {
        let today_total = self.get_today_total();
        let progress = self.get_progress_bar(today_total, self.goal, 30);

        println!("\n{}", c(&"═".repeat(50), CYAN));
        println!("{}", c("💧 WATER COUNTER – Daily Hydration Tracker", &format!("{}{}", BRIGHT, CYAN)));
        println!("{}", c(&"═".repeat(50), CYAN));
        println!("  Today: {}ml / {}ml  {}", today_total, self.goal, progress);
        println!("{}", c(&"─".repeat(50), DIM));
        println!("  1. 💧 Add water intake");
        println!("  2. 📊 Show today's progress");
        println!("  3. 📅 Show history ({} days)", HISTORY_DAYS);
        println!("  4. 📈 Show statistics");
        println!("  5. 🎯 Set daily goal (current: {}ml)", self.goal);
        println!("  6. 📤 Import / Export data");
        println!("  7. 🗑️  Clear all data");
        println!("  0. 🚪 Exit");
        println!("{}", c(&"═".repeat(50), CYAN));
    }

    fn handle_import_export(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        println!("\n📤 IMPORT / EXPORT");
        println!("  1. Export data");
        println!("  2. Import data");
        println!("  0. Back");
        let choice = Self::ask("Your choice: ");

        match choice.as_str() {
            "1" => self.export_data(None)?,
            "2" => {
                let fp = Self::ask("Path to JSON file: ");
                if !fp.is_empty() {
                    self.import_data(&fp)?;
                }
            }
            _ => {}
        }
        Ok(())
    }

    fn run(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        println!("{}", c("\n💧 Water Counter – Daily Hydration Tracker", &format!("{}{}", BRIGHT, CYAN)));
        println!("{}", c("Stay hydrated, stay healthy!", DIM));

        loop {
            self.show_menu();
            let choice = Self::ask("Your choice: ");

            match choice.as_str() {
                "1" => {
                    match Self::ask_int("Amount in ml: ") {
                        Ok(amount) => {
                            if let Err(e) = self.add_entry(amount) {
                                eprintln!("{}", c(&format!("❌ {}", e), RED));
                            }
                        }
                        Err(_) => eprintln!("{}", c("❌ Please enter a number.", RED)),
                    }
                }
                "2" => self.show_today(),
                "3" => self.show_history(HISTORY_DAYS),
                "4" => self.show_stats(),
                "5" => {
                    match Self::ask_int("New daily goal (ml): ") {
                        Ok(goal) => {
                            if let Err(e) = self.set_goal(goal) {
                                eprintln!("{}", c(&format!("❌ {}", e), RED));
                            }
                        }
                        Err(_) => eprintln!("{}", c("❌ Please enter a number.", RED)),
                    }
                }
                "6" => {
                    if let Err(e) = self.handle_import_export() {
                        eprintln!("{}", c(&format!("❌ {}", e), RED));
                    }
                }
                "7" => {
                    if let Err(e) = self.clear_data() {
                        eprintln!("{}", c(&format!("❌ {}", e), RED));
                    }
                }
                "0" => {
                    println!("{}", c("👋 Stay hydrated! Goodbye!", CYAN));
                    break;
                }
                _ => eprintln!("{}", c("❌ Invalid choice.", RED)),
            }

            if choice != "0" {
                print!("\nPress Enter to continue...");
                io::stdout().flush().unwrap();
                let mut _dummy = String::new();
                io::stdin().read_line(&mut _dummy).unwrap();
            }
        }

        Ok(())
    }
}

// ─── Main ─────────────────────────────────────────────────────────────────────

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut app = WaterCounter::new()?;
    app.run()?;
    Ok(())
}
